File: Android App/ColonelFund/app/src/main/java/com/colonelfund/colonelfund/Member.java
package com.colonelfund.colonelfund;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.Serializable;

/**
 * Basic Member Object that holds all the information for a member.
 */
public class Member implements Serializable {
    private static final long serialVersionUID = 2628890352033932835L;
    private String userID;
    private String firstName;
    private String lastName;
    private String emailAddress;
    private String phoneNumber;

    private String username;
    private String profilePicURL;
    private String facebookID;
    private String googleID;
    private String firebaseID;

    /**
     * Full constructor for member.
     *
     * @param userID of a member.
     * @param firstName of a member.
     * @param lastName of a member.
     * @param username of a member.
     * @param emailAddress of a member.
     * @param phoneNumber of a member.
     * @param profilePicURL of a member.
     * @param facebookID of a member.
     * @param googleID of a member.
     */
    public Member(String userID, String firstName, String lastName, String emailAddress, String phoneNumber, String username, String profilePicURL, String facebookID, String googleID, String firebaseID) {
        this.userID = userID;
        this.firstName = firstName;
        this.lastName = lastName;
        this.emailAddress = emailAddress;
        this.phoneNumber = phoneNumber;
        this.username = username;
        this.profilePicURL = profilePicURL;
        this.facebookID = facebookID;
        this.googleID = googleID;
        this.firebaseID = firebaseID;
        if (userID == "") {
            this.userID = "01"; //TODO: change once we have usernames figured out
        }

        if (username.equals("")) {
            makeUserName();
        }
    }

    /**
     * Constructor with just user id
     *
     * @param userID of a member.
     */
    public Member(String userID) {
        super();
        this.userID = userID;

        if (this.username.equals("")) {
            makeUserName();
        }
    }

    /**
     *
     * @return the firebaseID
     */
    public String getFirebaseID() {
        return firebaseID;
    }

    /**
     *
     * @param firebaseID the firebaseID to set
     */
    public void setFirebaseID(String firebaseID) {
        this.firebaseID = firebaseID;
    }

    /**
     * @return the userID
     */
    public String getUserID() {
        return userID;
    }

    /**
     * @param userID the userID to set
     */
    public void setUserID(String userID) {
        this.userID = userID;
    }

    /**
     * @return the firstName
     */
    public String getFirstName() {
        return firstName;
    }

    /**
     * @param firstName the firstName to set
     */
    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    /**
     * @return the lastName
     */
    public String getLastName() {
        return lastName;
    }

    /**
     * @return a formatted version of the member name. FirstName + LastName with
     * proper capitalization and a space between the names.
     */
    public String getFormattedFullName() {
        return (firstName.substring(0, 1).toUpperCase() + firstName.substring(1) +
                " " + lastName.substring(0, 1).toUpperCase() + lastName.substring(1));
    }

    /**
     * @param lastName the lastName to set
     */
    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    /**
     * @return the username
     */
    public String getUsername() {
        return username;
    }

    /**
     * @param username the username to set
     */
    public void setUsername(String username) {
        this.username = username;
    }

    /**
     * @return the emailAddress
     */
    public String getEmailAddress() {
        return emailAddress;
    }

    /**
     * @param emailAddress the emailAddress to set
     */
    public void setEmailAddress(String emailAddress) {
        this.emailAddress = emailAddress;
    }

    /**
     * @return the phoneNumber
     */
    public String getPhoneNumber() {
        return phoneNumber;
    }

    /**
     * @param phoneNumber the phoneNumber to set
     */
    public void setPhoneNumber(String phoneNumber) {
        this.phoneNumber = phoneNumber;
    }

    /**
     * @return the profilePicURL
     */
    public String getProfilePicURL() {
        return profilePicURL;
    }

    /**
     * @param profilePicURL the profilePicURL to set
     */
    public void setProfilePicURL(String profilePicURL) {
        this.profilePicURL = profilePicURL;
    }

    /**
     * @return the facebookID
     */
    public String getFacebookID() {
        return facebookID;
    }

    /**
     * @param facebookID the facebookID to set
     */
    public void setFacebookID(String facebookID) {
        this.facebookID = facebookID;
    }

    /**
     * @return the googleID
     */
    public String getGoogleID() {
        return googleID;
    }

    /**
     * @param googleID the googleID to set
     */
    public void setGoogleID(String googleID) {
        this.googleID = googleID;
    }

    /**
     * Creates a user name from first/last
     */
    private void makeUserName() {
        this.username = this.firstName.toLowerCase() + this.lastName.toLowerCase();
    }

    /**
     * Get a JSONObject representation of Member
     *
     * @return JSONObject
     */
    public JSONObject toJson() throws JSONException {
        JSONObject jsonObj = new JSONObject();
        jsonObj.put("userID", userID);
        jsonObj.put("firstName", firstName);
        jsonObj.put("lastName", lastName);
        jsonObj.put("emailAddress", emailAddress);
        jsonObj.put("phoneNumber", phoneNumber);

        jsonObj.put("username", username);
        jsonObj.put("profilePicURL", profilePicURL);
        jsonObj.put("facebookID", facebookID);
        jsonObj.put("googleID", googleID);
        jsonObj.put("firebaseID", firebaseID);
        return jsonObj;
    }

    /**
     * Constructor with JSON Object.
     *
     * @param jsonObject
     */
    public Member(JSONObject jsonObject) throws JSONException {
        this.userID = jsonObject.getString("userID");
        this.firstName = jsonObject.getString("firstName");
        this.lastName = jsonObject.getString("lastName");
        this.emailAddress = jsonObject.getString("emailAddress");
        this.phoneNumber = jsonObject.getString("phoneNumber");
        this.username = jsonObject.getString("username");
        this.profilePicURL = jsonObject.getString("profilePicURL");
        this.facebookID = jsonObject.getString("facebookID");
        this.googleID = jsonObject.getString("googleID");
        this.firebaseID = jsonObject.getString("firebaseID");
    }

}

File: iOS App/ColonelFund/ColonelFund/Member.swift
//
//  Member.swift
//  ColonelFund
//
//  Created by Chris Penner on 1/30/18.
//  Copyright © 2018 PennerTech. All rights reserved.
//

import Foundation

class Member: NSObject, Codable {
    private var userID: String
    private var firstName: String
    private var lastName: String
    private var username: String
    private var emailAddress: String
    private var phoneNumber: String
    private var profilePicURL: String
    private var facebookID: String //TODO: consider this for linking account
    private var googleID: String   //TODO: consider this for linking account
    private var associatedEvents: [Event]
    
    init(userID: String) {
        self.userID = userID
        self.firstName = ""
        self.lastName = ""
        self.username = ""
        self.emailAddress = ""
        self.phoneNumber = ""
        self.profilePicURL = ""
        self.facebookID = ""
        self.googleID = ""
        self.associatedEvents = []
    }
    
    init(userID: String, firstName: String, lastName: String, userName: String, emailAddress: String, phoneNumber: String) {
        self.userID = userID
        self.firstName = firstName
        self.lastName = lastName
        self.username = userName
        self.emailAddress = emailAddress
        self.phoneNumber = phoneNumber
        self.profilePicURL = ""
        self.facebookID = ""
        self.googleID = ""
        self.associatedEvents = []
    }
    
    init(facebookID: String, firstName: String, lastName: String, profilePicURL: String) {
        self.userID = ""
        self.firstName = firstName
        self.lastName = lastName
        self.username = ""
        self.emailAddress = ""
        self.phoneNumber = ""
        self.profilePicURL = profilePicURL
        self.facebookID = facebookID
        self.googleID = ""
        self.associatedEvents = []
        super.init()
        
        makeUsername()
        self.userID = "01" //TODO: change once we have usernames figured out
    }
    
    init(googleID: String, emailAddress: String, firstName: String, lastName: String, profilePicURL: String) {
        self.userID = ""
        self.firstName = firstName
        self.lastName = lastName
        self.username = ""
        self.emailAddress = emailAddress
        self.phoneNumber = ""
        self.profilePicURL = profilePicURL
        self.facebookID = ""
        self.googleID = googleID
        self.associatedEvents = []
        super.init()
        
        makeUsername()
        self.userID = "01" //TODO: change once we have usernames figured out
    }
    
    required init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        userID = try values.decode(String.self, forKey: .userID)
        firstName = try values.decode(String.self, forKey: .firstName)
        lastName = try values.decode(String.self, forKey: .lastName)
        emailAddress = try values.decode(String.self, forKey: .emailAddress)
        phoneNumber = try values.decode(String.self, forKey: .phoneNumber)
        username = try values.decode(String.self, forKey: .username)
        profilePicURL = try values.decode(String.self, forKey: .profilePicURL)
        facebookID = try values.decode(String.self, forKey: .facebookID)
        googleID = try values.decode(String.self, forKey: .googleID)
        self.username = firstName.lowercased() + lastName.lowercased()
        self.associatedEvents = []
    }
    
    func getUserID() -> String {
        return self.userID
    }
    
    func setUserID(userID: String) {
        self.userID = userID
    }
    
    func getFirstName() -> String {
        return self.firstName
    }
    
    func setFirstName(firstName: String) {
        self.firstName = firstName
    }
    
    func getLastName() -> String {
        return self.lastName
    }
    
    func setLastName(lastName: String) {
        self.lastName = lastName
    }
    
    func getUsername() -> String {
        return self.username
    }
    
    func setUsername(username: String) {
        self.username = username
    }
    
    func getEmailAddress() -> String {
        return self.emailAddress
    }
    
    func setEmailAddress(emailAddress: String) {
        self.emailAddress = emailAddress
    }
    
    func getPhoneNumber() -> String {
        return self.phoneNumber
    }
    
    func setPhoneNumber(phoneNumber: String) {
        self.phoneNumber = phoneNumber
    }
    
    func getProfilePicURL() -> String {
        return self.profilePicURL
    }
    
    func setProfilePicURL(profilePicURL: String) {
        self.profilePicURL = profilePicURL
    }
    
    func getFacebookID() -> String {
        return facebookID
    }
    
    func setFacebookID(facebookID: String) {
        self.facebookID = facebookID
    }
    
    func getAssociatedEvents() -> [Event] {
        return self.associatedEvents
    }
    
    func setAssociatedEvents() {
        self.associatedEvents.removeAll()
        for (item) in EventCollection.sharedInstance.eventArray {
            if (item.getAssociatedMember() == self.userID) {
                associatedEvents.append(item)
            }
        }
    }
    
    func getFormattedFullName() -> String {
        return self.firstName + " " + self.lastName
    }
    
    func makeUsername() {
        self.username = self.firstName.lowercased() + self.lastName.lowercased()
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(userID, forKey: .userID)
        try container.encode(firstName, forKey: .firstName)
        try container.encode(lastName, forKey: .lastName)
        try container.encode(emailAddress, forKey: .emailAddress)
        try container.encode(phoneNumber, forKey: .phoneNumber)
        try container.encode(username, forKey: .username)
        try container.encode(profilePicURL, forKey: .profilePicURL)
        try container.encode(facebookID, forKey: .facebookID)
        try container.encode(googleID, forKey: .googleID)
    }
    
    enum CodingKeys: String, CodingKey {
        case userID
        case firstName
        case lastName
        case username
        case emailAddress
        case phoneNumber
        case profilePicURL
        case facebookID
        case googleID
        case associatedEvents
    }
}
